File: Python3/206.Reverse_Linked_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

#iterative Solution
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev = None
        curr = head
        next_n = None
        while curr:
            next_n = curr.next
            curr.next = prev
            prev = curr
            curr = next_n
        return prev

#recursive Solution
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if (not head or not head.next):
            return head
        tail = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return tail

File: Java/206.Reverse_Linked_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
// iterative Solution
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        ListNode next = null;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }       
        return prev;
    }
}

// recursive Solution
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head; 
        ListNode tail = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return tail;
    }
}